### PR TITLE
Fixed registration asyncValidators and added tests

### PR DESCRIPTION
**Pull request by FrauDeborah**
Hi Johnathan,

I believe you can merge this PR to dev. 

This PR on devRefactorRegistration will complete the issues #7 Create a popup Register User Diaolg, and #64 RefactorRegistrationto fix Aysync Validator uses. Or at least I think the work is not complete. 

1) While writing tests for the registration, I did covered that the asyncValidators emailTaken and usernameTaken were improperly implements. 

- added EmailTakenAsyncValidatorDirective and the associated validator function with calls the emailTakenAsycValidator function. The emailTakenAsyncValidor function calls the registerService.isUniqueEmail.

- added test for EmailTakenAsyncValidatorDirective 
- added UsernameTakenAsyncValidatorDirective and the associated validator function with calls the usernameTakenAsycValidator function. The usernameTakenAsyncValidor function calls the registerService.isUniqueUsername.

- added test for UsernameTakenAsyncValidatorDirective 

2) added all test for RegisterComponent, and RegisterService

3) added (key.enter)="somefuntioncall... " to each registration buttons 

4) recall we decide earlier not to make registration a popup. 

Any feedback is appreciated or merge and close issue #7 and #64.

Thank you,

Deborah

### CODE REVIEW

File: src/app/registration/registration.component.spec.ts [hunk 1]
```diff
@@ -44,4 +50,202 @@ describe('RegistrationComponent', () => {
...
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.setValue("FirstnameNoSpaceLastname");
+    errors = fullnameFormControl.errors || {};
+    expect(errors['pattern']).toBeTruthy(); // has errors
+  });
+
+  // the following email field tests are to verify validaters email required and asyncValidator emailTaken work
+
+  it('should validate required email field value is present ', () => {
+    let errors = {};
```
Comment (haxwell) [inline src/app/registration/registration.component.spec.ts:107]: I think there is a bit of a cut and paste error in this test.. I think it is not validating the value of the email field as you intended..

Comment (haxwell) [reply src/app/registration/registration.component.spec.ts:107]: I didn't check all the tests, but you should, to see if there are any other similar errors, or tests that are not doing what you intended..

Comment (FrauDeborah) [reply src/app/registration/registration.component.spec.ts:107]: Johnathan (@haxwell) 

Wow, MyBad!!! I apologize for not catching the copy/paste errors previously. Shame on me for wasting your time. I think all those errors have been fixed now. Or at least I hope so ...

I also merged the recent dev change into this branch/PR.

Deborah

Comment (haxwell) [reply src/app/registration/registration.component.spec.ts:107]: No problem! I did not consider it a waste of time..

**Pull request merged**